src/components/tl/ucp/tl_ucp_context.c

```diff
@@ -392,30 +315,36 @@ ucc_status_t ucc_tl_ucp_ctx_remote_populate(ucc_tl_ucp_context_t * ctx,
                  MAX_NR_SEGMENTS);
         return UCC_ERR_INVALID_PARAM;
     }
-    remote_info = (ucc_tl_ucp_remote_info_t **)ucc_malloc(
-        sizeof(ucc_tl_ucp_remote_info_t *) * size, "ucp_ctx_remote_info");
-    if (NULL == remote_info) {
+    ctx->rkeys =
+        (ucp_rkey_h **)ucc_calloc(sizeof(ucp_rkey_h *), size, "ucp_ctx_rkeys");
+    if (NULL == ctx->rkeys) {
         tl_error(ctx->super.super.lib, "failed to allocated %zu bytes",
-                 sizeof(ucc_tl_ucp_remote_info_t *) * size);
+                 sizeof(ucp_rkey_h *) * size);
         return UCC_ERR_NO_MEMORY;
     }
-
-    remote_info[rank] = (ucc_tl_ucp_remote_info_t *)ucc_calloc(
+    for (i = 0; i < size; i++) {
+        ctx->rkeys[i] =
+            (ucp_rkey_h *)ucc_calloc(nsegs, sizeof(ucp_rkey_h), "rkey_per_seg");
```

Comment:
why to use double memory allocation for rkeys? Why not allocating rkys = malloc(size * nsegs * sizeof(ucp_rkey_h)) and then use an access macro, smth like
#define TL_UCP_REMOTE_RKEY(_ctx, _rank, _seg) (_ctx)->rkeys[ _seg * _ctx->size + _rank]